File: mobile/modules/herdr-ssh/android/src/main/java/dev/herdr/ssh/HerdrSshModule.kt
// Not from orca.
//
// ⚠️⚠️ **UNVERIFIED SOURCE.** This file has never been compiled. No Gradle sync, no `expo prebuild`,
// no APK, no device run — the receipt for this unit stops at the JavaScript boundary
// (`modules/herdr-ssh/test/`). The sshj symbols below were taken from its published API, but that is
// not the same as a green `assembleDebug`.
//
// The Android half of `NativeHerdrSsh` (`../../../../../../src/native-types.ts`). Obligations 4, 6,
// 7, 8 and 9 of `HerdrTransport` (`packages/herdr-client-ts/src/transport.ts`) are enforced in
// TypeScript above this file, where they are testable; what is owned here is 1 (no pty), 2 (the
// command string is passed through, never rebuilt), 3 (stderr rides the close) and 5 (one
// connection, N channels).
package dev.herdr.ssh

import expo.modules.kotlin.jni.JavaScriptFunction
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

/** `NativeSshConnectConfig` (`../src/native-types.ts`). */
class HerdrSshConnectConfig : Record {
  @Field var host: String = ""
  @Field var port: Int = 22
  @Field var username: String = ""
  @Field var privateKey: String = ""
  @Field var passphrase: String? = null

  /** Base64 body of the server's OpenSSH `SHA256:` fingerprint. */
  @Field var hostKeySha256: String? = null

  @Field var allowUnknownHostKey: Boolean = false
  @Field var connectTimeoutMs: Int = 20000
  @Field var keepaliveIntervalMs: Int = 15000
}

class HerdrSshModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("HerdrSsh")

    AsyncFunction("connect") { config: HerdrSshConnectConfig ->
      HerdrSshConnection(HerdrSshDialer.dial(config), appContext)
    }

    Class(HerdrSshConnection::class) {
      // The handlers are arguments, not listeners attached afterwards — obligation 4. sshj hands
      // back a `Session.Command` whose streams are already live, so a listener registered after
      // `openChannel` resolved would lose whatever the remote wrote first, which for a bridge that
      // failed to start is the entire diagnosis.
      AsyncFunction("openChannel") { connection: HerdrSshConnection,
        command: String,
        onData: JavaScriptFunction<Unit>,
        onClose: JavaScriptFunction<Unit> ->
        connection.openChannel(command, onData, onClose)
      }

      Function("disconnect") { connection: HerdrSshConnection ->
        connection.disconnect()
      }
    }

    Class(HerdrSshChannel::class) {
      // `ByteArray` in, `Uint8Array` on the JS side, with one `getRegion` copy and no base64:
      // `ByteArrayTypeConverter` declares `CppType.UINT8_TYPED_ARRAY` inbound, and outbound the
      // emit path routes through `FollyDynamicExtensionConverter` into `createUint8Array`
      // (expo-modules-core android/src/main/cpp/types/JNIToJSIConverter.cpp:16-41).
      AsyncFunction("write") { channel: HerdrSshChannel, bytes: ByteArray ->
        channel.write(bytes)
      }

      Function("close") { channel: HerdrSshChannel ->
        channel.close()
      }
    }
  }
}

File: mobile/modules/herdr-ssh/ios/HerdrSshModule.swift
// Not from orca.
//
// ⚠️⚠️ **UNVERIFIED SOURCE.** This file has never been compiled. No `expo prebuild`, no `pod
// install`, no Xcode build, no device run — the unit whose receipt this is stops at the JavaScript
// boundary (`modules/herdr-ssh/test/`). Every Citadel/NIOSSH symbol below was read out of the
// upstream sources at the pinned tags (see README.md §iOS) rather than typed from memory, but read
// is not compiled. Treat the whole file as a first draft that the first `pod install` will correct.
//
// What it is: the iOS half of `NativeHerdrSsh` (`../src/native-types.ts`). The nine obligations of
// `HerdrTransport` (`packages/herdr-client-ts/src/transport.ts`) are split deliberately — the five
// that are not about ssh (4, 6, 7, 8, 9) are enforced in TypeScript, above this file, where they are
// testable. What is left here is genuinely ssh, and the ones this file owns are:
//
//   1. **no pty** — `SSHClient.withExec` is the pty-less exec (`withPTY`/`withTTY` are its siblings
//      and are not used). There is no pty parameter on this module's API to get wrong.
//   2. **the command string** — `command` arrives from JS and is passed to `withExec` verbatim.
//      Nothing here parses, quotes or rebuilds it.
//   3. **stderr** — `ExecCommandOutput.stderr` is accumulated and delivered in the close payload.
//   5. **one connection, N channels** — one `SSHClient` per `HerdrSshConnection`; `openChannel`
//      opens a child channel on it and never dials.
import ExpoModulesCore
import Citadel
import Crypto
import Foundation
import NIOCore
import NIOSSH

/// `NativeSshConnectConfig` (`../src/native-types.ts`), as an Expo record.
struct HerdrSshConnectConfig: Record {
  @Field var host: String = ""
  @Field var port: Int = 22
  @Field var username: String = ""
  @Field var privateKey: String = ""
  @Field var passphrase: String?
  /// Base64 body of the server's OpenSSH `SHA256:` fingerprint.
  @Field var hostKeySha256: String?
  @Field var allowUnknownHostKey: Bool = false
  @Field var connectTimeoutMs: Int = 20000
  @Field var keepaliveIntervalMs: Int = 15000
}

public final class HerdrSshModule: Module {
  public func definition() -> ModuleDefinition {
    Name("HerdrSsh")

    AsyncFunction("connect") { (config: HerdrSshConnectConfig) -> HerdrSshConnection in
      let client = try await HerdrSshDialer.dial(config)
      return HerdrSshConnection(client: client)
    }

    Class(HerdrSshConnection.self) {
      // The handlers are *arguments*, not listeners added afterwards. That is obligation 4 and it
      // is the whole reason this is not an `EventEmitter`: the remote writes the instant the command
      // starts, and a listener attached after `openChannel` resolved would miss whatever preceded
      // it — including the entire stderr of a bridge that failed to start.
      AsyncFunction("openChannel") { (
        connection: HerdrSshConnection,
        command: String,
        onData: JavaScriptFunction<Void>,
        onClose: JavaScriptFunction<Void>
      ) -> HerdrSshChannel in
        let channel = HerdrSshChannel(onData: onData, onClose: onClose)
        channel.appContext = self.appContext
        try await connection.start(channel: channel, command: command)
        return channel
      }

      Function("disconnect") { (connection: HerdrSshConnection) in
        connection.disconnect()
      }
    }

    Class(HerdrSshChannel.self) {
      // `Data` in, `Uint8Array` on the JS side: `DynamicDataType` casts a `Uint8Array` straight into
      // `Data` over the JSI buffer (expo-modules-core/ios/Core/DynamicTypes/DynamicDataType.swift:20-25),
      // so a 55 KB frame crosses as one copy and never as base64.
      AsyncFunction("write") { (channel: HerdrSshChannel, bytes: Data) in
        try await channel.write(bytes)
      }

      Function("close") { (channel: HerdrSshChannel) in
        channel.close()
      }
    }
  }
}
